### ISSUE TITLE
kotlin-metadata class properties are always sorted alphabetically rather than in order of declaration

### ISSUE THREAD

**Issue by dimsuz** [square/kotlinpoet#965]
Having this class

```kotlin
data class Person(
  val firstName: String,
  val lastName: String,
  val age: String,
  val address: String,
  val phoneNumbers: List<String>,
  val friends: Map<String, String>
)
```
and calling
```kotlin
Person::class.toImmutableKmClass().properties.map { it.name }
```
produces:
```kotlin
[address, age, firstName, friends, lastName, phoneNumbers]
```

Is it possible to iterate them in order of declaration? 
Not sure if this is a `kotlin-metadata` or `kotlinpoet` issue.

**Comment by ZacSweers** [comment]
I don't remember if we sort them (we do sort some things) or if this is how they come out of metadata. Agree they should use declaration order though, PR welcome if you want to fix!

**Comment by dimsuz** [comment]
Yes, I'll try to research this once I have some time (if no one will beat me to it).

**Comment by dimsuz** [comment]
I've researched this a bit and this appears to be the order in which `kotlinx-metadata` calls its visitors.

In kotlinx-metadata there's [this call] with `propertyList` argument. `visitDeclarations` then calls `kotlinpoet`'s visitor with all these properties. And first one is already `address`, i.e sorted alphabetically.

I wanted to find how exactly that `propertyList` property is created. It's a member of `ProtoBuf.Class`, but then I get lost in a huge `JetBrains/kotlin` repo and failed to find this class. I don't want to download and import all that repo rn.

Maybe there is a way to get a declaration order through other means and then do additional sorting on `kotlinpoet`'s side? If anyone has a hint on how to do that, I might proceed, otherwise I think it should be reported to JetBrains. Youtrack I guess?

**Comment by ZacSweers** [comment]
Yeah youtrack is the best place to start and link it here

**Comment by dimsuz** [comment]
Reported to youtrack: https://youtrack.jetbrains.com/issue/KT-41042

**Comment by ZacSweers** [comment]
One thing we could look into in the mean time is trying to interpret the ordering from the reflection/elements source. Moshi would definitely want this. @swankjesse @Egorand what do you think?

**Comment by dimsuz** [comment]
I have just done that in my library. In my case that's an annotation processor, so I've used `TypeElement.enclosedElements` to read the field order and then sort kotlinpoet's property specs according to that.

**Comment by ZacSweers** [comment]
Want to send a PR? I think we need to factor method order too and give
priority in order of constructor parameters, public setters, public
getters, and finally public fields. I don’t think there’s a case where
there would be private anything that we’d use.

**Comment by Egorand** [comment]
Makes sense to me, a PR would be great!

**Comment by dimsuz** [comment]
I will try to come up with a PR, but not sure how fast I will able to do this. If anyone would like to work on this too, please do it.

**Comment by ZacSweers** [comment]
Just to share from the above youtrack link - this is an issue in metadata itself. kotlinx-serialization wants this ordering support though too, so optimistic something will come down the pipeline
